beets/util/__init__.py

```diff
@@ -188,28 +189,28 @@ def ancestry(path: bytes) -> List[str]:
 
 
 def sorted_walk(
-    path: AnyStr,
-    ignore: Sequence = (),
+    path: BytesOrStr,
```

Comment:
I was wondering why should paths be represented as `bytes` internally? This requirement may be less relevant now that python 2 is dropped and not supported anymore, see https://beets.io/blog/paths.html

> This way, Python 3’s Unicode [str](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str) can represent arbitrary bytes in filenames. (The first commit to beets happened a bit before Python 3.0 was released, so perhaps the project can be forgiven for not adopting this approach in the first place.)

Reply:
Ideally, I'd imagine these functions should accept the same types that the original ones do. In this case it's `os.walk`, which accepts any path-like `str` or `bytes` and returns _the same type_ that it received.

On the other hand, this PR was about fixing types rather than adjusting the implementation, so I focused on underlining what the current implement does and handles to make life easier to the person who will need to refactor this module later.

Reply:
> I was wondering why should paths be represented as bytes internally? This requirement may be less relevant now that python 2 is dropped and not supported anymore, see https://beets.io/blog/paths.html

Yes, that could be changed in principle, but that's in fact not completely straightforward: We still need conversions, at a minimum when dealing with the database (which currently stores bytes, not text. Would SQLite even support storing text with surrogate escapes?).
At that point, we might want to swtich to pathlib, however (in which case it's still not completely obvious what we should store to the database).

Reply:
SQLite stores unicode fine these days (we may want to include its icu extension to cover users that have some ancient versions of SQLite).

For example, I locally store bandcamp releases in this db:

```sql
CREATE TABLE release (
    source TEXT,
    label TEXT,
    name TEXT,
    album TEXT,
    artist TEXT,
    title TEXT,
    comment TEXT,
    url TEXT,
    email_id TEXT,
    added TEXT,
    done INTEGER DEFAULT 0,
    done_date TEXT,
    bought INTEGER DEFAULT 0,
    next INTEGER DEFAULT 0,
    invalid INTEGER DEFAULT 0,
    PRIMARY KEY(email_id, added, source, name) ON CONFLICT IGNORE
  );
``` 

See some of the data that I've got there 
![image](https://github.com/beetbox/beets/assets/16212750/066f8bdc-3cd1-48b1-8e13-abe21b910c48)

Reply:
On the other hand I'd strongly suggest to keep the database as it is, at least for now, since this means opening a can of worms of migrating existing user databases.

What I meant by 'internally', I guess, is the code handling the paths before it hits the database (I assume that regardless of whether we supply a `str` or `bytes`, db code appropriately converts it to `bytes` for storage).

On the other hand paths being stored as `bytes` in db doesn't mean that we need to handle them as such across the codebase. I agree that `pathlib` is indeed the right way forward.